feat: show space constrained UIs conditionally

kdheepak opened the pull request:
With this PR the constraint explorer demo only shows space constrained UIs instead:

Smallest (15 row height):

<img width="759" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/37a4a027-6c6d-4feb-8104-d732aee298ac">

Small (20 row height):

<img width="759" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/f76e025f-0061-4f09-9c91-2f7b00fcfb9e">

Medium (30 row height):

<img width="758" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/81b070da-1bfb-40c5-9fbc-c1ab44ce422e">

Full (40 row height):

<img width="760" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/7bb8a8c4-1a77-4bbc-a346-c8b5c198c6d3">

joshka commented:
I'd much rather see this collapse rather than hide the rows. The data presented could be reasonably useful in as few as 11 lines (or even 5 if pressed). Consider the following example conveys the same info (assuming colors make the distinction between blocks / spacers and that losing 1-2 columns for the labels is fine):
```
  ▌        Length(20)        ▐▌        Length(20)         ▐▌        Length(20)        ▐
  <---------------------------------------85 px--------------------------------------->
S:▌       20 px      ▐▌       20 px      ▐▌       20 px      ▐          25 px
C:    13 px    ▌       20 px      ▐▌       20 px      ▐▌       20 px      ▐   12 px
E:          25 px          ▌       20 px      ▐▌       20 px      ▐▌       20 px      ▐
A: 6 px ▌       20 px      ▐ 7 px  ▌       20 px      ▐ 6 px ▌       20 px      ▐ 6 px
B:▌       20 px      ▐    13 px    ▌       20 px      ▐   12 px    ▌       20 px      ▐
```
compared to:
```
▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜
▌        Length(20)        ▐▌        Length(20)         ▐▌        Length(20)        ▐
▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟
Flex::Start
<---------------------------------------85 px--------------------------------------->
▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜┌                       ┐
▌    Length(20)    ▐▌    Length(20)    ▐▌    Length(20)    ▐         Spacer
▌       20 px      ▐▌       20 px      ▐▌       20 px      ▐          25 px
▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟└                       ┘

Flex::Center
<---------------------------------------85 px--------------------------------------->
┌           ┐▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜┌          ┐
   Spacer    ▌    Length(20)    ▐▌    Length(20)    ▐▌    Length(20)    ▐   Spacer
    13 px    ▌       20 px      ▐▌       20 px      ▐▌       20 px      ▐   12 px
└           ┘▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟└          ┘

Flex::End
<---------------------------------------85 px--------------------------------------->
┌                       ┐▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜
         Spacer          ▌    Length(20)    ▐▌    Length(20)    ▐▌    Length(20)    ▐
          25 px          ▌       20 px      ▐▌       20 px      ▐▌       20 px      ▐
└                       ┘▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟

Flex::SpaceAround
<---------------------------------------85 px--------------------------------------->
┌    ┐▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜┌     ┐▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜┌    ┐▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜┌    ┐
Spacer▌    Length(20)    ▐Spacer ▌    Length(20)    ▐Spacer▌    Length(20)    ▐Spacer
 6 px ▌       20 px      ▐ 7 px  ▌       20 px      ▐ 6 px ▌       20 px      ▐ 6 px
└    ┘▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟└     ┘▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟└    ┘▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟└    ┘

Flex::SpaceBetween
<---------------------------------------85 px--------------------------------------->
▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜┌           ┐▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜┌          ┐▛▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▜
▌    Length(20)    ▐   Spacer    ▌    Length(20)    ▐   Spacer   ▌    Length(20)    ▐
▌       20 px      ▐    13 px    ▌       20 px      ▐   12 px    ▌       20 px      ▐
▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟└           ┘▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟└          ┘▙▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▟
```

Valentin271 commented:
Agree with joshka here I'd much prefer to find a way to shorten the content than hide it.

What about adding scroll? It would hide some flex combination but at least everything would be present.

The problem with hiding is that sometimes you can't make your terminal higher depending on font, screen size ...

joshka reviewed (approved):
LGTM, will push a fix to DRY the repetition in the methods before merging.

kdheepak commented:
The full demo displays completely fine in 40 row height. Zoomed in, I usually never go below 45 row. I'd be surprised if a user has a smaller than 40 px height terminal and wants to compare all the flex modes using this interactive demo. Showing a user 1 of the `Flex` modes with all the information present imo is better than constrained UI that remove useful info.

But I think designing space constrained UI is an interesting challenge and this _might_ be a good example of how to show blocks in 1px, 2px, 3px and 4px heights, so I've updated it that way. In the space constrained UIs, no labels are displayed but users can still do a full comparison.

I think we can do better if we conditionally opt for different layouts (e.g. `if area.height < 10 { Layout::vertical(...) } else { Layout::vertical(...) } `, which I didn't do. I just changed the blocks and spacers display based on available space. This results in some weird intermediate steps like this:

<img width="757" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/aa0e0ac3-2b0b-40e3-8169-f1f1a23eeee9">

Where some rows have 3 height and others have 2 height. Unfortunately, with the layout we have right now, this is the best we can do without conditionally opting in for different layouts.

joshka commented:
I'm going to push it without the fix I mentioned - I see the special casing is necessary and getting the right combination of colors with reuse is hard.

I wouldn't mind coming back to the idea mentioned here:

>I think we can do better if we conditionally opt for different layouts (e.g. if area.height < 10 { Layout::vertical([a, c]) } else { Layout::vertical([a, b, c, d]) }), which I didn't do. I just changed how the blocks and spacers render based on available space.
